File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/data/types/Location.java
/**
 * Copyright (c) Codice Foundation
 *
 * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
 * Lesser General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
 * License is distributed along with this program and can be found at
 * <http://www.gnu.org/licenses/lgpl.html>.
 */
package ddf.catalog.data.types;

/**
 * <b> This code is experimental. While this interface is functional and tested, it may change or be
 * removed in a future version of the library. </b>
 */
public interface Location {

  /**
   * {@link ddf.catalog.data.Attribute} name for accessing the altitude of the {@link
   * ddf.catalog.resource.Resource} of the {@link Metacard}. <br>
   */
  String ALTITUDE = "location.altitude-meters";

  /**
   * {@link ddf.catalog.data.Attribute} name for accessing the country code of the {@link
   * ddf.catalog.resource.Resource} of the {@link Metacard}. <br>
   */
  String COUNTRY_CODE = "location.country-code";

  /**
   * {@link ddf.catalog.data.Attribute} name for accessing the Coordinate Reference System (CRS) of
   * the {@link ddf.catalog.resource.Resource} of the {@link Metacard}. <br>
   */
  String COORDINATE_REFERENCE_SYSTEM_CODE = "location.crs-code";

  /**
   * {@link ddf.catalog.data.Attribute} name for accessing the CRS name of the {@link
   * ddf.catalog.resource.Resource} of the {@link Metacard}. <br>
   */
  String COORDINATE_REFERENCE_SYSTEM_NAME = "location.crs-name";
}

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/location/location.js
/**
 * Copyright (c) Codice Foundation
 *
 * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
 * is distributed along with this program and can be found at
 * <http://www.gnu.org/licenses/lgpl.html>.
 *
 **/
const React = require('react')

const CustomElements = require('../../js/CustomElements.js')

const Button = require('../button')
const Dropdown = require('../dropdown')
const { Menu, MenuItem } = require('../menu')
import styled from 'styled-components'

const Line = require('./line')
const Polygon = require('./polygon')
const PointRadius = require('./point-radius')
const BoundingBox = require('./bounding-box')
const Keyword = require('./keyword')
const plugin = require('plugins/location')

const inputs = plugin({
  line: {
    label: 'Line',
    Component: Line,
  },
  poly: {
    label: 'Polygon',
    Component: Polygon,
  },
  circle: {
    label: 'Point-Radius',
    Component: PointRadius,
  },
  bbox: {
    label: 'Bounding Box',
    Component: BoundingBox,
  },
  keyword: {
    label: 'Keyword',
    Component: ({ setState, keywordValue, ...props }) => {
      return (
        <Keyword
          {...props}
          value={keywordValue}
          setState={({ value, ...data }) => {
            setState({ keywordValue: value, ...data })
          }}
          setBufferState={(key, value) => setState({ [key]: value })}
        />
      )
    },
  },
})

const drawTypes = ['line', 'poly', 'circle', 'bbox']

const Form = ({ children }) => (
  <div className="form-group clearfix">{children}</div>
)

const DrawButton = ({ onDraw }) => (
  <Button className="location-draw is-primary" onClick={onDraw}>
    <span className="fa fa-globe" />
    <span>Draw</span>
  </Button>
)

const Root = styled.div`
  height: ${props => (props.isOpen ? 'auto' : props.theme.minimumButtonSize)};
`

const Component = CustomElements.registerReact('location')
const LocationInput = props => {
  const { mode, setState } = props
  const input = inputs[mode] || {}
  const { Component: Input = null } = input
  return (
    <Root isOpen={input.label !== undefined}>
      <Component>
        <Dropdown label={input.label || 'Select Location Option'}>
          <Menu value={mode} onChange={value => setState({ ['mode']: value })}>
            {Object.keys(inputs).map(key => (
              <MenuItem key={key} value={key}>
                {inputs[key].label}
              </MenuItem>
            ))}
          </Menu>
        </Dropdown>
        <Form>
          {Input !== null ? <Input {...props} /> : null}
          {drawTypes.includes(mode) ? (
            <DrawButton onDraw={props.onDraw} />
          ) : null}
        </Form>
      </Component>
    </Root>
  )
}

module.exports = ({ state, setState, options }) => (
  <LocationInput {...state} onDraw={options.onDraw} setState={setState} />
)
